File: my-code/213/main.py
# -*- coding:utf-8 -*-
# @Time     : 2021/05/28 18: 02
# @Author   : Ranshi
# @File     : 213. 打家劫舍 II.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        def robRange(start: int, end: int) -> int:
            first = nums[start]
            second = max(nums[start], nums[start + 1])
            for i in range(start + 2, end + 1):
                first, second = second, max(first + nums[i], second)
            return second

        length = len(nums)
        if length == 1:
            return nums[0]
        elif length == 2:
            return max(nums[0], nums[1])
        else:
            return max(robRange(0, length - 2), robRange(1, length - 1))


if __name__ == '__main__':
    s = Solution()
    print(s.rob(nums=[6, 6, 4, 8, 4, 3, 3, 10]))

File: my-code/213/main.go
package main

import "fmt"

// @Time     : 2021/08/08 14:49
// @Author   : Ranshi
// @File     : main.go
func maxInt(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func rob(nums []int) int {
	if len(nums) == 1 {
		return nums[0]
	} else if len(nums) == 2 {
		return maxInt(nums[0], nums[1])
	} else if len(nums) == 3 {
		return maxInt(maxInt(nums[0], nums[1]), nums[2])
	}
	maxMoney := func(arr []int) int {
		if len(arr) == 1 {
			return arr[0]
		}
		dp := make([]int, len(arr))
		dp[0] = arr[0]
		dp[1] = maxInt(arr[0], arr[1])
		for i := 2; i < len(arr); i++ {
			dp[i] = maxInt(dp[i-1], dp[i-2]+arr[i])
		}
		return dp[len(dp)-1]
	}
	return maxInt(maxMoney(nums[1:]), maxMoney(nums[2:len(nums)-1])+nums[0])
}

func main() {
	fmt.Println(rob([]int{0}))
}
